app/src/main/java/com/lagradost/cloudstream3/plugins/PluginManager.kt

```diff
@@ -265,16 +267,102 @@ object PluginManager {
         }
 
         main {
-            createNotification(activity, updatedPlugins)
+            createNotification(activity, updatedPlugins, R.string.plugins_updated)
         }
 
-       // ioSafe {
+        // ioSafe {
             afterPluginsLoadedEvent.invoke(true)
-       // }
+        // }
 
         Log.i(TAG, "Plugin update done!")
     }
 
+    /**
+     * Automatically download plugins not yet existing on local
+     * 1. Gets all online data from online plugins repo
+     * 2. Fetch all not downloaded plugins
+     * 3. Download them and reload plugins
+     **/
+    fun downloadNotExistingPluginsAndLoad(activity: Activity, isLoaded: Boolean) {
+        // Load all plugins as fast as possible, if its not loaded yet.
+        if (!isLoaded) {
+            loadAllOnlinePlugins(activity)
+            afterPluginsLoadedEvent.invoke(true)
+        }
+
+        val newDownloadPlugins = mutableListOf<String>()
+        val urls = (getKey<Array<RepositoryData>>(REPOSITORIES_KEY)
+            ?: emptyArray()) + PREBUILT_REPOSITORIES
+        val onlinePlugins = urls.toList().apmap {
+            getRepoPlugins(it.url)?.toList() ?: emptyList()
+        }.flatten().distinctBy { it.second.url }
+
+        val providerLang = activity.getApiProviderLangSettings()
+        //Log.i(TAG, "providerLang => ${providerLang.toJson()}")
+
+        // Iterate online repos and returns not downloaded plugins
+        val notDownloadedPlugins = onlinePlugins.mapNotNull outer@{ onlineData ->
+            val sitePlugin = onlineData.second
+            //Don't include empty urls
+            if (sitePlugin.url.isBlank()) { return@outer null }
+            if (sitePlugin.repositoryUrl.isNullOrBlank()) { return@outer null }
+
+            //Omit already existing plugins
+            if (getPluginPath(activity, sitePlugin.internalName, onlineData.first).exists()) {
+                Log.i(TAG, "Skip > ${sitePlugin.internalName}")
+                return@outer null
+            }
+
+            //Omit lang not selected on language setting
+            val lang = sitePlugin.language ?: return@outer null
+            //If set to 'universal', don't skip any language
+            if (!providerLang.contains(AllLanguagesName) && !providerLang.contains(lang)) {
+                return@outer null
+            }
+            //Log.i(TAG, "sitePlugin lang => $lang")
+
+            //Omit NSFW, if disabled
+            sitePlugin.tvTypes?.let { tvtypes ->
+                if (!settingsForProvider.enableAdult) {
+                    if (tvtypes.contains("NSFW")) {
```

Comment:
I can see why you did this as we do have another reference in the codebase doing this, but I'd rather have TvType.NSFW.name for some kind of future proofing :pensive: